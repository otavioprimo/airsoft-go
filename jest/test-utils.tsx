import React from 'react';
import {render} from '@testing-library/react-native';
import {ThemeProvider} from 'styled-components/native';
import {createStore, Reducer} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';

import * as Theme from '../src/theme';

interface ReduxProviders {
  state: any;
  reducer: Reducer<any, any>;
}

interface RenderProviders {
  children: React.ReactElement;
  options?: ReduxProviders;
}

const ReduxProvider = ({children, options}: any) => {
  if (!options) {
    return <>{children}</>;
  }

  const store = createStore(options.reducer, options.state);
  return <Provider store={store}>{children}</Provider>;
};

const AllTheProviders: React.FC<RenderProviders> = ({children, options}) => {
  return (
    <ThemeProvider theme={Theme.DarkTheme}>
      <ReduxProvider options={options}>{children}</ReduxProvider>
    </ThemeProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: ReduxProviders) => {
  return render(<AllTheProviders options={options}>{ui}</AllTheProviders>);
};

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
