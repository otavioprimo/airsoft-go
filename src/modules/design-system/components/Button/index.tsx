import React from 'react';
import {TouchableOpacityProps} from 'react-native';

import {
  Link,
  Basic,
  Outline,
  LabelLink,
  LabelBasic,
  LabelOutline,
  LoadingLink,
  LoadingBasic,
  LoadingOutline,
} from './style';

type VariantValues = 'basic' | 'link' | 'outline';

interface ButtonProps extends TouchableOpacityProps {
  label: string;
  ref?: any;
  variant?: VariantValues;
  full?: boolean;
  loading?: boolean;
  onlyLoading?: boolean;
}

interface VariantComponents {
  basic: any;
  outline: any;
  link: any;
}

const buttons: VariantComponents = {
  basic: Basic,
  outline: Outline,
  link: Link,
};

const labels: VariantComponents = {
  basic: LabelBasic,
  outline: LabelOutline,
  link: LabelLink,
};

const loadings: VariantComponents = {
  basic: LoadingBasic,
  outline: LoadingOutline,
  link: LoadingLink,
};

const Button: React.FC<ButtonProps> = React.forwardRef(
  (props: ButtonProps, ref: any) => {
    let {variant} = props;
    if (!variant) {
      variant = 'basic';
    }

    const Component = buttons[variant];
    const LabelComponent = labels[variant];
    const Loading = loadings[variant];

    return (
      <Component {...props} full={props.full} ref={ref}>
        {props.loading && <Loading />}
        <LabelComponent disabled={props.disabled}>{props.label}</LabelComponent>
      </Component>
    );
  },
);

export default React.memo(Button);
