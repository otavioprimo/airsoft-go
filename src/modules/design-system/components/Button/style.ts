import styled from 'styled-components/native';

interface ButtonProps {
  full?: boolean;
}

interface LabelProps {
  disabled?: boolean;
}

const Button = styled.TouchableOpacity`
  width: ${({full}: ButtonProps) => (full ? '100%' : '200px')};
  height: 50px;
  background-color: ${props =>
    props.disabled ? props.theme.disabled : props.theme.secondary};
  margin-bottom: 8px;
  align-items: center;
  justify-content: center;
  flex-direction: row;
  border-radius: 4px;
`;

export const Basic = styled(Button)``;

export const Outline = styled(Button)`
  background-color: transparent;
  border-width: 1px;
  border-color: ${props =>
    props.disabled ? props.theme.disabled : props.theme.secondary};
`;

export const Link = styled(Button)`
  background-color: transparent;
`;

const Label = styled.Text`
  font-size: 16px;
  font-weight: 700;
`;

export const LabelBasic = styled(Label)`
  color: ${({theme}) => theme.textSecondary};
`;

export const LabelOutline = styled(Label)<LabelProps>`
  color: ${props =>
    props.disabled ? props.theme.disabled : props.theme.textPrimary};
`;

export const LabelLink = styled(Label)<LabelProps>`
  color: ${props =>
    props.disabled ? props.theme.disabled : props.theme.textPrimary};
`;

const Loading = styled.ActivityIndicator.attrs(({theme}: any) => ({
  size: 'small',
  color: theme.secondary,
}))`
  margin-right: 8px;
`;

export const LoadingBasic = styled(Loading).attrs(({theme}) => ({
  color: theme.primary,
}))``;

export const LoadingOutline = styled(Loading)``;

export const LoadingLink = styled(Loading)``;
