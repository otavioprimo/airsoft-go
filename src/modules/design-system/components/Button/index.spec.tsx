import React from 'react';
import {act} from 'react-test-renderer';
import {render} from 'test/test-utils';

import Button from './index';

afterEach(async () => {
  await act(async () => {});
});

describe('Design System - Button', () => {
  it('Should match snapshot - default', () => {
    const {toJSON} = render(<Button label="Sign-up" />);
    expect(toJSON()).toMatchSnapshot('Button_default');
  });

  it('Should match snapshot - basic', () => {
    const {toJSON} = render(<Button label="Sign-up" variant="basic" />);
    expect(toJSON()).toMatchSnapshot('Button_basic');
  });

  it('Should match snapshot - outline', () => {
    const {toJSON} = render(<Button label="Sign-up" variant="outline" />);
    expect(toJSON()).toMatchSnapshot('Button_outline');
  });

  it('Should match snapshot - link', () => {
    const {toJSON} = render(<Button label="Sign-up" variant="link" />);
    expect(toJSON()).toMatchSnapshot('Button_link');
  });
});
