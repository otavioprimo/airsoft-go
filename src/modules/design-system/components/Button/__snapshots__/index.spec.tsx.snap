// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design System - Button Should match snapshot - basic: Button_basic 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(230,230,230,.9)",
      "borderRadius": 4,
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "marginBottom": 8,
      "opacity": 1,
      "width": 100,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#222",
          "fontSize": 16,
          "fontWeight": "700",
          "marginLeft": 8,
        },
      ]
    }
  >
    Sign-up
  </Text>
</View>
`;

exports[`Design System - Button Should match snapshot - default: Button_default 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(230,230,230,.9)",
      "borderRadius": 4,
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "marginBottom": 8,
      "opacity": 1,
      "width": 100,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#222",
          "fontSize": 16,
          "fontWeight": "700",
          "marginLeft": 8,
        },
      ]
    }
  >
    Sign-up
  </Text>
</View>
`;

exports[`Design System - Button Should match snapshot - link: Button_link 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderRadius": 4,
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "marginBottom": 8,
      "opacity": 1,
      "width": 100,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#fff",
          "fontSize": 16,
          "fontWeight": "700",
          "marginLeft": 8,
        },
      ]
    }
  >
    Sign-up
  </Text>
</View>
`;

exports[`Design System - Button Should match snapshot - outline: Button_outline 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderColor": "rgba(230,230,230,.9)",
      "borderRadius": 4,
      "borderWidth": 1,
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "marginBottom": 8,
      "opacity": 1,
      "width": 100,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#fff",
          "fontSize": 16,
          "fontWeight": "700",
          "marginLeft": 8,
        },
      ]
    }
  >
    Sign-up
  </Text>
</View>
`;
